lib/Doctrine/ORM/Persisters/ManyToManyPersister.php

```diff
@@ -77,7 +77,19 @@ public function update(PersistentCollection $coll)
      */
     public function get(PersistentCollection $coll, $index)
     {
-        throw new \BadMethodCallException("Selecting a collection by index is not supported by this CollectionPersister.");
+        $mapping = $coll->getMapping();
+
+        if ( ! isset($mapping['indexBy'])) {
+            throw new \BadMethodCallException("Selecting a collection by index is only supported on indexed collections.");
```

Comment:
Not part of this scope... others does the same, keeping the standard.
Dedicated Exception should be a separate PR changing everything.